File: Python/ByTier/A/2052_A_Adrenaline_Rush.py
# Problem: CF 2052 A - Adrenaline Rush
# https://codeforces.com/contest/2052/problem/A

"""
Algorithm/Techniques: Greedy Algorithm, Set Operations

Time Complexity: O(n^2)
Space Complexity: O(n^2)

This code computes the maximum number of overtakes that could have occurred during a race,
given the final order of cars. It simulates a process where cars are inserted one by one
into a sequence and each new car overtakes all cars that come after it in the current list.
The result is a list of overtakes in temporal order, such that when applied to an initial
sorted sequence, it produces the given final order.

The greedy strategy ensures all possible overtakes occur, respecting the constraint that
a car cannot overtake another more than once. The solution builds the sequence step by step,
tracking which cars have already overtaken others and ensuring no duplicate overtakes.
"""

n = int(input())
c = list(map(int, input().split()))
ans = []

# For each car in the final order, simulate insertions
for i in range(n):
    for j in range(i + 1, n):
        # Add overtakes in reverse order: j overtakes i (j is behind i)
        ans.append(f"{j+1} {i+1}")

s = set()
# Process each car in the final order
for i in range(n):
    x = c[i]
    s.add(x)
    # For all positions after this car's current position, simulate overtakes
    for j in range(x + 1, n + 1):
        if j not in s:
            # Car x overtakes car j if j was not inserted yet
            ans.append(f"{x} {j}")

print(len(ans))
print("\n".join(ans))


# https://github.com/VaHiX/codeForces/

File: CPP/ByRound/2052/2052_A_Adrenaline_Rush.cpp
// Problem: CF 2052 A - Adrenaline Rush
// https://codeforces.com/contest/2052/problem/A

/*
Purpose: This code determines the maximum number of overtakes that could have occurred 
         during a race to achieve a given final car order. It simulates the process 
         of reversing the final order into the initial order by swapping adjacent cars,
         counting each swap as an overtake event.

Algorithms/Techniques: 
- Simulation technique to reverse the final order to initial order.
- Greedy swapping approach to minimize the number of swaps needed.
- Uses a mapping from car positions in final order to determine the necessary swaps.

Time Complexity: O(n^2) where n is the number of cars, due to nested loops for finding 
                 the correct position and swapping.

Space Complexity: O(n) for storing car positions and the sequence of overtakes.
*/

#include <stddef.h>
#include <algorithm>
#include <iostream>
#include <map>
#include <utility>
#include <vector>

#define ll long long
#define uint unsigned int
using namespace std;
constexpr int m = 998244353;

void call() {
  int n;
  cin >> n;
  vector<int> v(n), ans(n);  // v stores final order, ans stores current order
  map<int, int> m;  // Maps a car to the car it overtakes
  vector<pair<int, int>> sol;  // Stores the sequence of overtakes
  for (int i = 0; i < n; i++) {
    cin >> v[i];
    ans[i] = i + 1;  // Initialize ans to [1, 2, ..., n]
  }
  // Build mapping of which car overtakes which based on final order
  for (int i = 0, j = 0; i < n; i++) {
    for (j = i; j < n; j++) {
      if (v[i] > v[j]) {  // If v[i] should be ahead of v[j] in final arrangement
        m[v[i]] = v[j];  // v[i] overtakes v[j]
        break;
      }
    }
    if (j == n)  // No car after i is smaller than v[i]
      m[v[i]] = -1;
  }
  int k = 0;
  // Simulate the process until the order matches final arrangement
  while (k < n) {
    for (int i = 1; i < n; i++) {
      // If the car at position i-1 should overtake the car at position i
      if (m[ans[i - 1]] == ans[i]) {
        break;  // No more overtakes needed at this step
      }
      // Record the overtake event: ans[i-1] overtakes ans[i]
      sol.push_back({ans[i], ans[i - 1]});
      // Swap the cars in the current order
      int t = ans[i - 1];
      ans[i - 1] = ans[i];
      ans[i] = t;
    }
    k++;
  }
  cout << sol.size() << "\n";  // Print total number of overtakes
  for (auto x : sol)  // Output all overtakes
    cout << x.first << " " << x.second << "\n";
}

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  int t = 1;
  while (t--) {
    call();
    cout << endl;
  }
}


// https://github.com/VaHiX/CodeForces/